File: app/models/issue_activity.rb
# == Schema Information
#
# Table name: issue_activities
#
#  id           :bigint           not null, primary key
#  issue_id     :integer
#  member_id    :integer
#  property     :string
#  before_value :string
#  after_value  :string
#  created_at   :datetime         not null
#  updated_at   :datetime         not null
#

class IssueActivity < ApplicationRecord
  belongs_to :issue
  belongs_to :member

  [
    [ 'category', 'category_id', "Category" ],
    [ 'creator', 'creator_id', "Member" ],
    [ 'assignee', 'assignee_id', "Member" ],
    [ 'milestone', 'milestone_id', "Milestone" ],
    [ 'project', 'project_id', "Project" ],
  ].each do |code, property, class_name|
    belongs_to :"before_#{code}", ->(record) { none if record.property != property }, foreign_key: :before_value, class_name: class_name, optional: true
    belongs_to :"after_#{code}", ->(record) { none if record.property != property }, foreign_key: :after_value, class_name: class_name, optional: true
  end

  cleanup_column :before_value, :after_value
end

File: app/javascript/frontend/models/IssueActivity.ts
import { Type } from "class-transformer"
import { Member } from "./Member"
import * as t from '@/lib/transforms'
import { Category } from "./Category"
import { Milestone } from "./Milestone"
import { Project } from "./Project"
import type { IssueActivitySchema } from './schema/issue_activity'

export class IssueActivity implements IssueActivitySchema {
  @t.Number id!: number
  @t.Number issueId!: number
  @t.Number memberId!: number
  @t.String property!: string
  @t.String afterValue: string | null = null
  @t.String beforeValue: string | null = null

  @t.Klass(Member) member?: Member

  @t.Date createdAt!: Date
  @t.Date updatedAt!: Date

  @t.Klass(Category) beforeCategory?: Category
  @t.Klass(Category) afterCategory?: Category

  @t.Klass(Member) beforeCreator?: Member
  @t.Klass(Member) afterCreator?: Member

  @t.Klass(Member) beforeAssignee?: Member
  @t.Klass(Member) afterAssignee?: Member

  @t.Klass(Milestone) beforeMilestone?: Milestone
  @t.Klass(Milestone) afterMilestone?: Milestone

  @t.Klass(Project) beforeProject?: Project
  @t.Klass(Project) afterProject?: Project
}
